File: java/spring-boot/src/main/java/com/backendtemplate/repositories/ReservationRepository.java
package com.backendtemplate.repositories;

import com.backendtemplate.contracts.Reservation;
import com.backendtemplate.exceptions.IdempotencyClaimed;
import jakarta.persistence.EntityManager;
import java.util.Optional;
import org.hibernate.exception.ConstraintViolationException;
import org.springframework.context.annotation.Profile;
import org.springframework.stereotype.Repository;

@Repository
@Profile("!no-db")
public class ReservationRepository {
    private final EntityManager entities;
    private final ProductRepository products;
    private final ReservationReplayRepository replays;

    public ReservationRepository(EntityManager entities, ProductRepository products, ReservationReplayRepository replays) {
        this.entities = entities;
        this.products = products;
        this.replays = replays;
    }

    public void claim(String key) {
        // Assigned IDs must INSERT, never merge an existing claim into a successful no-op.
        entities.persist(new ReservationClaimEntity(key));
        try {
            entities.flush();
        } catch (ConstraintViolationException failure) {
            if (isH2ClaimInsertConflict(failure)) {
                throw new IdempotencyClaimed();
            }
            throw failure;
        }
    }

    private static boolean isH2ClaimInsertConflict(ConstraintViolationException failure) {
        // The claim table has only one unique constraint: its idempotency-key primary key.
        return "23505".equals(failure.getSQLState()) && failure.getSQL() != null
                && failure.getSQL().contains("insert into reservation_claims");
    }

    public Optional<Reservation> findReplay(String key) {
        return replays.findByKey(key).map(ReservationReplayEntity::toContract);
    }

    public boolean decreaseStockIfAvailable(String productId) {
        return products.decreaseAvailableStock(productId) > 0;
    }

    public boolean productExists(String productId) {
        return products.existsById(productId);
    }

    public void saveReservationAndReplay(Reservation reservation, String key) {
        var product = entities.getReference(ProductEntity.class, reservation.productId());
        var stored = new ReservationEntity(reservation, product);
        entities.persist(stored);
        entities.persist(new ReservationReplayEntity(key, reservation, stored));
        // Flush is not completion; the Service proxy still owns commit and any rollback.
        entities.flush();
    }

}

File: kotlin/spring-boot/src/main/kotlin/com/backendtemplate/repositories/ReservationRepository.kt
package com.backendtemplate.repositories

import com.backendtemplate.contracts.Reservation
import com.backendtemplate.exceptions.IdempotencyClaimed
import jakarta.persistence.EntityManager
import org.hibernate.exception.ConstraintViolationException
import org.springframework.context.annotation.Profile
import org.springframework.stereotype.Repository

@Repository
@Profile("!no-db")
class ReservationRepository(
    private val entities: EntityManager,
    private val products: ProductRepository,
    private val replays: ReservationReplayRepository,
) {
    fun claim(key: String) {
        entities.persist(ReservationClaimEntity(key))
        try {
            entities.flush()
        } catch (failure: ConstraintViolationException) {
            if (isH2ClaimInsertConflict(failure)) throw IdempotencyClaimed()
            throw failure
        }
    }

    fun findReplay(key: String): Reservation? = replays.findByKey(key)?.toContract()

    fun decreaseStockIfAvailable(productId: String): Boolean = products.decreaseAvailableStock(productId) > 0

    fun productExists(productId: String): Boolean = products.existsById(productId)

    fun saveReservationAndReplay(reservation: Reservation, key: String) {
        val product = entities.getReference(ProductEntity::class.java, reservation.productId)
        val stored = ReservationEntity(reservation, product)
        entities.persist(stored)
        entities.persist(ReservationReplayEntity(key = key, value = reservation, reservation = stored))
        entities.flush()
    }
}

private fun isH2ClaimInsertConflict(failure: ConstraintViolationException): Boolean =
    failure.sqlState == "23505" && failure.sql?.contains("insert into reservation_claims") == true
